File: proj/hwvc_android/codec_native/src/main/java/com/lmy/hwvcnative/processor/AlImageProcessor.kt
package com.lmy.hwvcnative.processor

import android.graphics.PointF
import android.os.Handler
import android.os.Looper
import android.view.Surface
import androidx.annotation.Keep
import com.lmy.hwvcnative.CPPObject
import com.lmy.hwvcnative.entity.AlLayer
import com.lmy.hwvcnative.entity.AlRational
import com.lmy.hwvcnative.entity.AlResult

@Keep
class AlImageProcessor private constructor() : CPPObject() {
    private val mMainHandler = Handler(Looper.getMainLooper())
    private var onSaveListener: OnSaveListener? = null
    private var onLayerInfoListener: OnLayerInfoListener? = null

    init {
        handler = create()
    }

    fun release() {
        if (!isNativeNull()) {
            release(handler)
        }
        handler = 0L
    }

    fun updateWindow(surface: Surface) {
        if (!isNativeNull()) {
            updateWindow(handler, surface)
        }
        invalidate()
    }

    fun invalidate() {
        if (!isNativeNull()) {
            invalidate(handler)
        }
    }

    /**
     * 设置画布大小，图层也会等比例缩放
     * @param w     新的画布宽度
     * @param h     新的画布高度
     * @param loc   定位
     * @param color 画布颜色
     */
    fun setCanvas(w: Int, h: Int, loc: Int = 0, color: Int = 0) {
        if (!isNativeNull()) {
            setCanvas(handler, w, h, loc, color)
        }
    }

    /**
     * 添加新的图层
     * @param path 图片文件路径
     * @return     新图层的id，id < 0表示添加失败，可用于后续的图层操作
     */
    fun addLayer(path: String): Int {
        if (!isNativeNull()) {
            return addLayer(handler, path)
        }
        return AlResult.FAILED
    }

    /**
     * 改变图层层级
     * @param id    图层id
     * @param index 图层层级，index >= 0，if(index > layer size) index = layer size
     *              0为最底层，会被所有图层遮挡
     * @return      {@link AlResult}
     */
    fun moveLayerIndex(id: Int, index: Int): Int {
        if (!isNativeNull()) {
            return moveLayerIndex(handler, id, index)
        }
        return AlResult.FAILED
    }

    /**
     * 移除图层
     * @param id 图层id
     * @return   {@link AlResult}
     */
    fun removeLayer(id: Int): Int {
        if (!isNativeNull()) {
            return removeLayer(handler, id)
        }
        return AlResult.FAILED
    }

    /**
     * 缩放图层
     * @param id    图层id
     * @param scale 缩放倍数
     * @return      {@link AlResult}
     */
    fun setScale(id: Int, scale: AlRational): Int {
        if (!isNativeNull()) {
            return setScale(handler, id, scale.num, scale.den)
        }
        return AlResult.FAILED
    }

    /**
     * 累乘缩放图层
     * @param id     图层id
     * @param ds     缩放因子
     * @param anchor 缩放锚点
     * @return       {@link AlResult}
     */
    fun postScale(id: Int, ds: AlRational, anchor: PointF): Int {
        if (!isNativeNull()) {
            return postScale(handler, id, ds.num, ds.den, anchor.x, anchor.y)
        }
        return AlResult.FAILED
    }

    /**
     * 旋转图层
     * @param id 图层id
     * @param r  旋转弧度，分数表示，没有乘以PI，顺时针方向为正向(CW)
     * @return   {@link AlResult}
     */
    fun setRotation(id: Int, r: AlRational): Int {
        if (!isNativeNull()) {
            return setRotation(handler, id, r.num, r.den)
        }
        return AlResult.FAILED
    }

    /**
     * 累加旋转图层
     * @param id     图层id
     * @param dr     旋转弧度，分数表示，没有乘以PI，顺时针方向为正向(CW)
     * @param anchor 旋转锚点
     * @return       {@link AlResult}
     */
    fun postRotation(id: Int, dr: AlRational, anchor: PointF): Int {
        if (!isNativeNull()) {
            return postRotation(handler, id, dr.num, dr.den, anchor.x, anchor.y)
        }
        return AlResult.FAILED
    }

    /**
     * 移动图层
     * @param id 图层id
     * @param x  相对于画布的x轴位移，画布中心为0点，x∈[-1, 1]为画布可见范围，绝对值大于1时会超出画布范围而不可见
     * @param y  相对于画布的y轴位移，同x
     * @return   {@link AlResult}
     */
    fun setTranslate(id: Int, x: Float, y: Float): Int {
        if (!isNativeNull()) {
            return setTranslate(handler, id, x, y)
        }
        return AlResult.FAILED
    }

    /**
     * 移动图层，会累加前一次的位移
     * @param id 图层id
     * @param dx  相对于画布的x轴位移，画布中心为0点
     * @param dy  相对于画布的y轴位移，同dx
     * @return   {@link AlResult}
     */
    fun postTranslate(id: Int, dx: Float, dy: Float): Int {
        if (!isNativeNull()) {
            return postTranslate(handler, id, dx, dy)
        }
        return AlResult.FAILED
    }

    /**
     * 设置图层透明度
     * @param id    图层id
     * @param alpha 图层透明度，范围[0, 1]，1为完全不透明
     * @return      {@link AlResult}
     */
    fun setAlpha(id: Int, alpha: Float): Int {
        if (!isNativeNull()) {
            return setAlpha(handler, id, alpha)
        }
        return AlResult.FAILED
    }

    /**
     * 根据View坐标获取layer id
     * 用于点击选择图层
     * @param x 相对于View的x轴坐标
     * @param y 相对于View的y轴坐标
     * @return  图层id，负数表示没找到
     */
    fun getLayer(x: Float, y: Float): Int {
        if (!isNativeNull()) {
            return getLayer(handler, x, y)
        }
        return AlResult.FAILED
    }

    /**
     * 对图层进行裁剪，目前一个图层只能有一个裁剪操作
     * 如果要更新裁剪区域，请先进行cancelCropLayer
     * 该操作会重置图层的缩放旋转位移
     * @param id      图层id
     * @param left   左, left∈[0, 1]
     * @param top    上, top∈[0, 1]
     * @param right  右, right∈[0, 1]
     * @param bottom 下, bottom∈[0, 1]
     * @return       {@link AlResult}
     */
    fun ensureCropLayer(id: Int, left: Float, top: Float, right: Float, bottom: Float): Int {
        if (!isNativeNull()) {
            return ensureCropLayer(handler, id, left, top, right, bottom)
        }
        return AlResult.FAILED
    }

    /**
     * 裁剪画布，并保持各图层在画布中的位置不变
     * @param left   左, left∈[-1, 1]
     * @param top    上, top∈[-1, 1]
     * @param right  右, right∈[-1, 1]
     * @param bottom 下, bottom∈[-1, 1]
     */
    fun cropCanvas(left: Float, top: Float, right: Float, bottom: Float) {
        if (!isNativeNull()) {
            cropCanvas(handler, left, top, right, bottom)
        }
    }

    /**
     * 取消图层裁剪
     * @param id      图层id
     * @return       {@link AlResult}
     */
    fun cancelCropLayer(id: Int): Int {
        if (!isNativeNull()) {
            return cancelCropLayer(handler, id)
        }
        return AlResult.FAILED
    }

    /**
     * 校直裁剪
     * @param id 图层id
     * @param r  旋转角度
     * @return   {@link AlResult}
     */
    fun ensureAlignCrop(id: Int, r: AlRational): Int {
        if (!isNativeNull()) {
            return ensureAlignCrop(handler, id, r.num, r.den)
        }
        return AlResult.FAILED
    }

    /**
     * 取消校直裁剪
     * @param id 图层id
     * @return   {@link AlResult}
     */
    fun cancelAlignCrop(id: Int): Int {
        if (!isNativeNull()) {
            return cancelAlignCrop(handler, id)
        }
        return AlResult.FAILED
    }

    /**
     * 保存图片
     * @return {@link AlResult}
     */
    fun save(path: String): Int {
        if (!isNativeNull()) {
            return save(handler, path)
        }
        return AlResult.FAILED
    }

    /**
     * 导出到.alx文件
     */
    fun export(path: String): Int {
        if (!isNativeNull()) {
            return export(handler, path)
        }
        return AlResult.FAILED
    }

    /**
     * 从.alx文件导入
     */
    fun import(path: String): Int {
        if (!isNativeNull()) {
            return import(handler, path)
        }
        return AlResult.FAILED
    }

    fun redo(): Int {
        if (!isNativeNull()) {
            return redo(handler)
        }
        return AlResult.FAILED
    }

    fun undo(): Int {
        if (!isNativeNull()) {
            return undo(handler)
        }
        return AlResult.FAILED
    }

    /**
     * 画笔绘制接口
     * @param point 屏幕像素坐标，中心为原点
     */
    fun paint(id: Int, point: PointF, painting: Boolean): Int {
        if (!isNativeNull()) {
            return paint(handler, id, point.x, point.y, painting)
        }
        return AlResult.FAILED
    }

    fun queryLayerInfo(l: OnLayerInfoListener? = null) {
        if (null != l) {
            onLayerInfoListener = l
        }
        if (!isNativeNull()) {
            return queryLayerInfo(handler)
        }
    }

    /***************************/
    /**      Listener         **/
    /***************************/
    interface OnSaveListener {
        fun onSave(code: Int, msg: String?, path: String?)
    }

    fun setOnSaveListener(listener: OnSaveListener) {
        this.onSaveListener = listener
    }

    fun setOnSaveListener(listener: (code: Int, msg: String?, path: String?) -> Unit) {
        setOnSaveListener(object : OnSaveListener {
            override fun onSave(code: Int, msg: String?, path: String?) {
                listener(code, msg, path)
            }
        })
    }

    interface OnLayerInfoListener {
        fun onInfo(layers: Array<AlLayer>)
    }

    fun setOnLayerInfoListener(l: OnLayerInfoListener) {
        this.onLayerInfoListener = l
    }

    fun setOnLayerInfoListener(l: (layers: Array<AlLayer>) -> Unit) {
        setOnLayerInfoListener(object : OnLayerInfoListener {
            override fun onInfo(layers: Array<AlLayer>) {
                l(layers)
            }
        })
    }

    /***************************/
    /** Callback from native  **/
    /***************************/

    /**
     * 图像保存完成回调
     * @param code {@link AlResult}
     * @param msg  状态信息
     * @param path 保存路径
     */
    fun onSave(code: Int, msg: String?, path: String?) {
        mMainHandler.post {
            onSaveListener?.onSave(code, msg, path)
        }
    }

    /**
     * 查询所有图层信息回调
     * @param ids layer id
     * @param ws  layer width
     * @param hs  layer height
     */
    fun onLayerInfo(ids: IntArray, ws: IntArray, hs: IntArray) {
        val layers = Array(ids.size) {
            return@Array AlLayer(ids[it], ws[it], hs[it])
        }
        mMainHandler.post {
            onLayerInfoListener?.onInfo(layers)
        }
    }


    companion object {
        fun create(): AlImageProcessor = AlImageProcessor()
    }

    private external fun create(): Long
    private external fun release(handler: Long)
    private external fun updateWindow(handler: Long, surface: Surface)
    private external fun invalidate(handler: Long)
    private external fun setCanvas(handler: Long, w: Int, h: Int, loc: Int, color: Int)
    private external fun addLayer(handler: Long, path: String): Int
    private external fun moveLayerIndex(handler: Long, id: Int, index: Int): Int
    private external fun removeLayer(handler: Long, id: Int): Int
    private external fun setScale(handler: Long, id: Int, scaleNum: Int, scaleDen: Int): Int
    private external fun postScale(handler: Long, id: Int, dsNum: Int, dsDen: Int,
                                   anchorX: Float, anchorY: Float): Int

    private external fun setRotation(handler: Long, id: Int, rNum: Int, rDen: Int): Int
    private external fun postRotation(handler: Long, id: Int, drNum: Int, drDen: Int,
                                      anchorX: Float, anchorY: Float): Int

    private external fun setTranslate(handler: Long, id: Int, x: Float, y: Float): Int
    private external fun postTranslate(handler: Long, id: Int, dx: Float, dy: Float): Int
    private external fun setAlpha(handler: Long, id: Int, alpha: Float): Int
    private external fun getLayer(handler: Long, x: Float, y: Float): Int
    private external fun ensureCropLayer(handler: Long, id: Int, left: Float, top: Float,
                                         right: Float, bottom: Float): Int

    private external fun cropCanvas(handler: Long, left: Float, top: Float,
                                    right: Float, bottom: Float)

    private external fun cancelCropLayer(handler: Long, id: Int): Int

    private external fun ensureAlignCrop(handler: Long, id: Int, rNum: Int, rDen: Int): Int

    private external fun cancelAlignCrop(handler: Long, id: Int): Int

    private external fun save(handler: Long, path: String): Int
    private external fun export(handler: Long, path: String): Int
    private external fun import(handler: Long, path: String): Int
    private external fun redo(handler: Long): Int
    private external fun undo(handler: Long): Int
    private external fun paint(handler: Long, id: Int, x: Float, y: Float, painting: Boolean): Int
    private external fun queryLayerInfo(handler: Long)
}

File: src/al_image/core/AlImageProcessor.cpp
/*
* Copyright (c) 2018-present, aliminabc@gmail.com.
*
* This source code is licensed under the MIT license found in the
* LICENSE file in the root directory of this source tree.
*/

#include "AlImageProcessor.h"
#include "AlUImageCodec.h"
#include "AlGImage.h"
#include "AlScreen.h"
#include "ObjectBox.h"
#include "AlLayerActionFactory.h"
#include "AlMCropAction.h"
#include "AlContext.h"
#include "AlCoordsTranslator.h"
#include "AlTarUtil.h"
#include "AlMath.h"
#include "AlRenderParams.h"
#include "AlPaintDesc.h"
#include "AlOperateScale.h"
#include "AlOperateRotate.h"
#include "AlOperateTrans.h"
#include "AlOperateAlpha.h"
#include "AlOperateCrop.h"
#include "AlFuture.h"

#define TAG "AlImageProcessor"

AlImageProcessor::AlImageProcessor() : AlAbsProcessor("AlImageProcessor") {
//    tar_t *archive = nullptr;
//    int count = tar_read_file("/sdcard/test.tar", &archive);
//    if (archive) {
//        for (int i = 0; i < count; ++i) {
//            Logcat::i(TAG, "%s", archive[i].name);
//        }
//        tar_free(archive);
//    }
    registerAnUnit(new AlScreen(ALIAS_OF_SCREEN));
    registerAnUnit(new AlUImageCodec(ALIAS_OF_IMAGE));
    registerAnUnit(new AlGImage(ALIAS_OF_IMAGE_GRAPH));
    registerEvent(EVENT_LAYER_MEASURE_CANVAS_NOTIFY,
                  reinterpret_cast<EventFunc>(&AlImageProcessor::_onCanvasUpdate));
    registerEvent(EVENT_LAYER_QUERY_ID_NOTIFY,
                  reinterpret_cast<EventFunc>(&AlImageProcessor::_onLayerQuery));
    registerEvent(EVENT_LAYER_EXPORT_FINISH,
                  reinterpret_cast<EventFunc>(&AlImageProcessor::_onExportFinish));
    registerEvent(EVENT_LAYER_IMPORT_FINISH,
                  reinterpret_cast<EventFunc>(&AlImageProcessor::_onImportFinish));
    registerEvent(EVENT_CANVAS_SAVE_FINISH,
                  reinterpret_cast<EventFunc>(&AlImageProcessor::_onSaveFinish));
    registerEvent(EVENT_LAYER_QUERY_INFO_NOTIFY,
                  reinterpret_cast<EventFunc>(&AlImageProcessor::_onLayerInfoResult));
}

AlImageProcessor::~AlImageProcessor() {
    this->onSaveListener = nullptr;
}

void AlImageProcessor::onCreate() {
    AlAbsProcessor::onCreate();
    Logcat::i(TAG, "%s(%d)", __FUNCTION__, __LINE__);
    this->aBaseCtx = AlEgl::offScreen(TAG);
    this->context = new AlContext();

    AlMessage *msg = AlMessage::obtain(MSG_CANVAS_SET_BG);
    msg->arg1 = static_cast<int32_t>(AlImageCanvas::kBGType::GRID);
    postEvent(msg);
//    this->putObject("AL_CONTEXT", this->context)
//            .to({ALIAS_OF_LAYER, ALIAS_OF_DESCRIPTOR, ALIAS_OF_CANVAS, ALIAS_OF_SCREEN});
}

void AlImageProcessor::onDestroy() {
    AlAbsProcessor::onDestroy();
    Logcat::i(TAG, "%s(%d)", __FUNCTION__, __LINE__);
    delete this->context;
    this->context = nullptr;
    delete this->aBaseCtx;
    this->aBaseCtx = nullptr;
}

void AlImageProcessor::updateWindow(HwWindow *win) {
    mWinSize.width = win->getWidth();
    mWinSize.height = win->getHeight();
    AlMessage *msg = AlMessage::obtain(EVENT_SCREEN_UPDATE_WINDOW, new NativeWindow(win, nullptr));
    postEvent(msg);
}

void AlImageProcessor::setCanvas(int32_t w, int32_t h, AlRectLoc loc, AlColor color) {
    mCanvasSize.width = w;
    mCanvasSize.height = h;
    auto *msg = AlMessage::obtain(MSG_LAYER_RESIZE_CANVAS,
                                  new AlRectLoc(loc));
    msg->arg1 = w;
    msg->arg2 = h;
    postEvent(msg);
}

void AlImageProcessor::invalidate(int32_t flag) {
    AlMessage *msg = AlMessage::obtain(EVENT_COMMON_INVALIDATE, nullptr,
                                       AlMessage::QUEUE_MODE_UNIQUE);
    msg->arg1 = flag;
    postEvent(msg);
}

int32_t AlImageProcessor::addLayer(const char *path) {
    AlMessage *msg = AlMessage::obtain(EVENT_LAYER_ADD);
    msg->arg1 = EVENT_LAYER_ADD;
    msg->desc = std::string(path);
    postEvent(msg);
    mQueryLock.wait();
    return mCurLayerId;
}

HwResult AlImageProcessor::removeLayer(int32_t id) {
    AlMessage *msg = AlMessage::obtain(EVENT_LAYER_REMOVE);
    msg->arg1 = id;
    postEvent(msg);
    return Hw::SUCCESS;
}

HwResult AlImageProcessor::moveLayerIndex(int32_t id, int32_t index) {
    AlMessage *msg = AlMessage::obtain(EVENT_LAYER_MOVE);
    msg->arg1 = id;
    msg->arg2 = index;
    postEvent(msg);
    return Hw::SUCCESS;
}

HwResult AlImageProcessor::setScale(int32_t id, AlRational scale) {
    auto *msg = AlMessage::obtain(EVENT_LAYER_SCALE,
                                  new AlOperateScale(id, scale, AlPointF(0, 0)));
    postEvent(msg);
    return Hw::SUCCESS;
}

HwResult AlImageProcessor::postScale(int32_t id, AlRational ds, AlPointF anchor) {
    auto *msg = AlMessage::obtain(EVENT_LAYER_SCALE_POST,
                                  new AlOperateScale(id, ds, anchor));
    postEvent(msg);
    return Hw::SUCCESS;
}

HwResult AlImageProcessor::setRotation(int32_t id, AlRational r) {
    auto *msg = AlMessage::obtain(EVENT_LAYER_ROTATE,
                                  new AlOperateRotate(id, r, AlPointF(0, 0)));
    postEvent(msg);
    return Hw::SUCCESS;
}

HwResult AlImageProcessor::postRotation(int32_t id, AlRational dr, AlPointF anchor) {
    auto *msg = AlMessage::obtain(EVENT_LAYER_ROTATE_POST,
                                  new AlOperateRotate(id, dr, anchor));
    postEvent(msg);
    return Hw::SUCCESS;
}

HwResult AlImageProcessor::setTranslate(int32_t id, float x, float y) {
    auto *msg = AlMessage::obtain(EVENT_LAYER_TRANS,
                                  new AlOperateTrans(id, x, y));
    postEvent(msg);
    return Hw::SUCCESS;
}

HwResult AlImageProcessor::postTranslate(int32_t id, float dx, float dy) {
    auto *msg = AlMessage::obtain(EVENT_LAYER_TRANS_POST,
                                  new AlOperateTrans(id, dx, dy));
    postEvent(msg);
    return Hw::SUCCESS;
}

HwResult AlImageProcessor::setAlpha(int32_t id, float alpha) {
    auto *msg = AlMessage::obtain(EVENT_LAYER_TRANS_POST,
                                  new AlOperateAlpha(id, alpha));
    postEvent(msg);
    return Hw::SUCCESS;
}

int32_t AlImageProcessor::getLayer(float x, float y) {
    auto bundle = std::make_shared<AlFuture>();
    auto *msg = AlMessage::obtain(MSG_LAYER_QUERY_ID_FUTURE,
                                  new AlOperateTrans(0, x, y));
    msg->sp = bundle;
    postEvent(msg);
    return bundle->get(-1);
}

HwResult AlImageProcessor::cropLayer(int32_t id, float left, float top, float right, float bottom) {
    auto *msg = AlMessage::obtain(EVENT_LAYER_CROP,
                                  new AlOperateCrop(id, left, top, right, bottom),
                                  AlMessage::QUEUE_MODE_UNIQUE);
    postEvent(msg);
    return Hw::SUCCESS;
}

HwResult AlImageProcessor::cropCanvas(float left, float top, float right, float bottom) {
    auto *msg = AlMessage::obtain(EVENT_CANVAS_CROP,
                                  new AlOperateCrop(AlIdentityCreator::NONE_ID,
                                                    left, top, right, bottom),
                                  AlMessage::QUEUE_MODE_UNIQUE);
    postEvent(msg);
    return Hw::SUCCESS;
}

HwResult AlImageProcessor::cancelCropLayer(int32_t id) {
    auto *msg = AlMessage::obtain(EVENT_LAYER_CROP_CANCEL,
                                  new AlOperateCrop(id, 0, 0, 0, 0),
                                  AlMessage::QUEUE_MODE_UNIQUE);
    postEvent(msg);
    return Hw::SUCCESS;
}

HwResult AlImageProcessor::save(std::string path) {
    auto *m = AlMessage::obtain(EVENT_CANVAS_SAVE);
    m->arg1 = EVENT_CANVAS_SAVE;
    m->desc = path;
    postMessage(m);
    AlRenderParams params;
    params.setRenderScreen(false);
    params.setTransparent(true);
    invalidate(params.toInt());
    return Hw::SUCCESS;
}

HwResult AlImageProcessor::exportFile(std::string path) {
    AlMessage *msg = AlMessage::obtain(EVENT_LAYER_EXPORT);
    msg->desc = path;
    postEvent(msg);
    mExportLock.wait();
    return Hw::SUCCESS;
}

HwResult AlImageProcessor::importFile(std::string path) {
    AlMessage *msg = AlMessage::obtain(EVENT_LAYER_IMPORT);
    msg->desc = path;
    postEvent(msg);
    invalidate();
    return Hw::SUCCESS;
}

void AlImageProcessor::setOnSaveListener(OnSaveListener listener) {
    this->onSaveListener = listener;
}

HwResult AlImageProcessor::ensureAlignCrop(int32_t id, AlRational r) {
    auto *msg = AlMessage::obtain(EVENT_LAYER_ALIGN_CROP,
                                  new AlOperateRotate(id, r, AlVec2(0, 0)),
                                  AlMessage::QUEUE_MODE_UNIQUE);
    postEvent(msg);
    return Hw::SUCCESS;
}

HwResult AlImageProcessor::cancelAlignCrop(int32_t id) {
    auto *msg = AlMessage::obtain(EVENT_LAYER_ALIGN_CROP_CANCEL,
                                  new AlOperateRotate(id, AlRational(), AlVec2()),
                                  AlMessage::QUEUE_MODE_UNIQUE);
    postEvent(msg);
    return Hw::SUCCESS;
}

HwResult AlImageProcessor::redo() {
    postEvent(AlMessage::obtain(EVENT_AIMAGE_REDO));
    return Hw::SUCCESS;
}

HwResult AlImageProcessor::undo() {
    postEvent(AlMessage::obtain(EVENT_AIMAGE_UNDO));
    return Hw::SUCCESS;
}

HwResult AlImageProcessor::paint(int32_t id, float x, float y, bool painting) {
    auto *msg = AlMessage::obtain(EVENT_LAYER_PAINT);
    msg->obj = new AlPaintDesc(id, x, y, painting);
    postEvent(msg);
    return Hw::SUCCESS;
}

void AlImageProcessor::queryLayerInfo() {
    postEvent(AlMessage::obtain(EVENT_LAYER_QUERY_INFO));
}

bool AlImageProcessor::_onCanvasUpdate(AlMessage *msg) {
    mCanvasSize.width = msg->arg1;
    mCanvasSize.height = static_cast<int>(msg->arg2);
    return true;
}

bool AlImageProcessor::_onLayerQuery(AlMessage *msg) {
    AlLogI(TAG, "%d", msg->arg1);
    mCurLayerId = msg->arg1;
    mQueryLock.notify();
    return true;
}

bool AlImageProcessor::_onExportFinish(AlMessage *msg) {
    mExportLock.notify();
    return true;
}

bool AlImageProcessor::_onImportFinish(AlMessage *msg) {
    return true;
}

bool AlImageProcessor::_onSaveFinish(AlMessage *msg) {
    if (this->onSaveListener) {
        int ret = static_cast<int>(msg->arg2);
        std::string tip("Save finish.");
        if (Hw::SUCCESS.code != ret) {
            tip = "Save failed.";
        }
        this->onSaveListener(ret, tip.c_str(), msg->desc.c_str());
    }
    return true;
}

bool AlImageProcessor::_onLayerInfoResult(AlMessage *msg) {
    auto *box = msg->getObj<ObjectBox *>();
    if (this->onLayerInfoListener && box) {
        auto *models = static_cast<vector<AlImgLayerDescription *> *>(box->ptr);
        std::vector<int32_t> ids(models->size());
        std::vector<int32_t> ws(models->size());
        std::vector<int32_t> hs(models->size());
        for (int i = 0; i < models->size(); ++i) {
            ids[i] = (*models)[i]->getId();
            AlSize size = (*models)[i]->getSize();
            ws[i] = size.width;
            hs[i] = size.height;
        }
        this->onLayerInfoListener(ids, ws, hs);
        delete models;
    }
    return true;
}

void AlImageProcessor::setOnLayerInfoListener(AlImageProcessor::OnLayerInfoListener l) {
    this->onLayerInfoListener = l;
}
